### ISSUE TITLE
Is there a limit to the 'Export' option?

### ISSUE THREAD

**Issue by logusergithub** [ssborbis/ContextSearch-web-ext#468]
After adding the latest sites, I clicked on 'Export'. The backup's popup has not appeared. Will it ever be that I exaggerated with search engines? I duplicated the Firefox profile and then deleted one of the main folders. The popup to export the backup has appeared.
I repeated the same procedure by deleting another folder to make sure the previous one was not the cause of the issue and I also checked with the previous extension version (1.35.2). Latest esported backup .json was 29,28 MB.

**Comment by ssborbis** [comment]
There shouldn't be a limit, unless you've hit some limit of the localStorage API. Can you open the web console from the CS options page and check for errors (Ctrl+Shift+K usually)

**Comment by ssborbis** [comment]
BTW, that's a pretty large export. Do you have a very large cache limit size for images?

**Comment by logusergithub** [comment]
On the console with CS and F5:

Source map error: Error: NetworkError when attempting to fetch resource.
Resource URL: moz-extension://b6eb6dea-e000-48ce-9466-a6b1fbf17c49/lib/browser-polyfill.min.js
Source map URL: browser-polyfill.min.js.map

I made a new profile (same error) and imported the last mentioned backup: with surprise no export popup, but... After I clicked on 'Delete search engine icon cache' the popup has appeared. So I tried on the main profile but this time I don't see nothing on the console when I clicked on the delete button.

"Do you have a very large cache limit size for images?"
You mean some parameter on about:config (with user.js 'browser.cache.disk.enable' is set to false') or CS 'cacheIconsMaxSize' (it is set to 256)?

**Comment by logusergithub** [comment]
I wanted to be sure that it could be replicated not to waste your time and with the export settings closest to the current one. Try this backup on a new profile: if you import it two times you should see that you are unable to export until you click on 'Delete search engine icon cache'.
I also took a look what happens in the extension's folder (profile -> storage -> default) when you import the backup the first and second time:  https://i.postimg.cc/7hXGtCDz/2022-03-02-122007.png

Note: currently on the main profile the 'Delete search engine icon cache' (iconUtils.js on the console) doesn't start, but he does it after I imported the backup I gave you. After that:
- reload CS options --> no backup's export
- after click on 'Delete icon cache' --> backup's export
- reload CS options -> no backup's export

https://privfile.com/download.php?fid=621f664ed6f9c-MTI1NjA=

Pass: howtobreakanextension! (a bit of a good mood about the mess I've made)

**Comment by ssborbis** [comment]
I just remembered, from the advanced menu, you can enable `exportWithoutBase64Icons`

This will effectively remove any icon cache for engines that have a valid icon URL before exporting. This should reduce the size of the export if that's an issue, but obviously, they will need to be reloaded & cached on import.

--

I think I did get the export to fail once, but I didn't see any errors in the console. I'm moving the export function to a blob to see if that helps at all. There is slightly revised code in the development branch if you wanted to test.

**Comment by logusergithub** [comment]
Same password above:
https://privfile.com/download.php?fid=621fe9e9ee9d0-MTI1NjU=

New profile:
- first import: export ok (storage profile folder file 26,78 MB)
- second import of the same backup: unable to export (storage profile folder file became 54,49 MB) and 'Delete icon cache'\'reload CS options' cycle
- exportWithoutBase64Icons: seems export always ok until you deselect it followed by 'reload CS options' (but it doesn't work on the main profile)

Note: caching icons 705\706.

**Comment by ssborbis** [comment]
Ok, I got your config. I'll load and test it later tonight

>Where I can find the link for the test?

You'd need to sideload as a temporary addon in FF from about:debugging, or an unpacked extension in Chrome from extensions + developer mode.

Hold up if you want until I have a chances to test your config.

**Comment by logusergithub** [comment]
Ok, thanks. Out of curiosity I tried but I receive a "does not contain a valid manifest". Since you have my configuration to check, I am reminded some points that may perhaps be useful to you:

- the selection icons in the search engine manager (audio\frame\image\link\page\seletion\video) are for speeding up instead of opening the settings one by one, right? I'm not sure they work instead of make the selection from the search engine window regarding the context menu

- due to the fact I use external 'data:image/x-icon;base64' I encounter an error when exporting to Firefox search bar. Not a big deal and I brought this on myself, but I note that help popup during the process keeps reappearing even if you select 'Don't show up it again'

- new app launcher setting does not store 'Modify Terms Regex'. I tried with /^$/{url}/g you gave me

**Comment by ssborbis** [comment]
I tried loading your config. It does import correctly, but fails to export IF I'm using the old code ( 1.36 or less ). BUT, if I use the latest code in the dev branch ( using blob ), it does export. I'm thinking the issue may be some limit to the amount of data a single element cant hold. I'd been putting the exported string into a `<a>` element, attached to the DOM. Because changing the export to type-`Blob` seems to fix it, that looks like the culprit. I'll leave it to you to test. I'll make a signed beta for you to try.

> the selection icons in the search engine manager (audio\frame\image\link\page\seletion\video) are for speeding up instead of opening the settings one by one, right?

Yes

> I'm not sure they work instead of make the selection from the search engine window regarding the context menu

It's working for me on the latest code. Be aware, if you have the config page open when you perform a search on another tab, the Options page can have a slightly outdated config, because the action of performing a search updates the user preferences objects and saves it, but the options page might not reflect the updated object in the forms / checkboxes / etc. 

Another possibility is there's some corruption in your CS config and it's not saving properly. I've seen this happen on Firefox due to DeadObject types that get caught in the JSON object.

> due to the fact I use external 'data:image/x-icon;base64' I encounter an error when exporting to Firefox search bar. Not a big deal and I brought this on myself, but I note that help popup during the process keeps reappearing even if you select 'Don't show up it again'

Hmm.. that would be an issue. I don't think the OpenSearch standard allows for data URIs. When you say "exporting to Firefox search bar", do you mean the button at the bottom of the edit forms "Install to Firefox One-Click Search Engines" or are you using Advanced -> Export to Firefox search?

**Comment by ssborbis** [comment]
Here's a beta. The only change is migrating the export to `Blob`. Again, I had no trouble importing your config and exporting using the new code, or any problem changing the contexts (audio/image/link/etc...) using the icons. There may be more going on with your setup, however.

[contextsearch_web_ext-1.37b1-an+fx.xpi.zip](https://github.com/ssborbis/ContextSearch-web-ext/files/8174239/contextsearch_web_ext-1.37b1-an%2Bfx.xpi.zip)

You'll need to unzip before installing the .xpi

**Comment by logusergithub** [comment]
A victory in the night! No issue with exporting. I imported it multiple times. 30,4 MB if I click on 'Delete search engine icon cache' first, 64,2 MB if I don't.

> or any problem changing the contexts (audio/image/link/etc...)

I recently tried the Context Menu after the external apps feature, so I used those fast buttons. Maybe I miss a piece but from the following image shouldn't the Context Menu must show all the engines selected with 'A' when I perform a selection. With those search engines not displayed I solved opening the related windows and deselecting\reselecting 'Selection' box.

<img src="https://i.postimg.cc/63mZS8kT/Search-For-Selection.png">

> "Install to Firefox One-Click Search Engines"

Yes. I noticed that if I import an .ico file to CS engine setting it will be converted to a png, moreover with a little bit increasing in size (I'm already pretty crazy to use 256x256). 

<img src="https://i.postimg.cc/PxxZ2WFB/Icons.png">

To preserving the the .ico file, I always use ACSE extension to convert the icon. Then I paste the code into CS 'Icon URI'. Maybe that's why I receive the metioned error if I try to export from CS the search engine to Firefox (obviously I don't if I use CS's icons selected or converted):

```
Non è stato possibile scaricare il plugin di ricerca da: https://opensearch-api.appspot.com/?SHORTNAME=Openclipart&DESCRIPTION=&TEMPLATE=https%3A%2F%2Fopenclipart.org%2Fsearch%2F%3Fquery%3D%257BsearchTerms%257D&POST_PARAMS=&METHOD=GET&ENCODING=UTF-8&ICON=data%3Aimage%2Fx-icon%3Bbase64%2CAAABAAEAAAAAAAEAGAD0LQAAFgAAAIlQTkcNChoKAAAADUlIRFIAAAEAAAABAAgGAAAAXHKoZgAAIABJREFUeNrtnXdAFUfXxp%2B9l0vvRaQqKN2Cotgjdo1dYxRjT6KABTUxaow1GttrTYw9xcSuFFGxYO8oiqCADRSlSe%2Ftlu%2BPJO%2Bb%2BMVEYPbe2WV%2BfyUCs2fOnnl2dnbmHM5ksbkKDAajXiJhLmAwmAAwGAwmAAwGgwkAg8FgAsBgMJgAMBgMJgAMBoMJAIPBYALAYDCYADAYDCYADAaDCQCDwWACwGAwmAAwGAwmAAwGgwkAg8GgDi0xdUZHSwd6Mj3oaOlAi9OCCkpUyqtQqahEeXU5FEoFu%2BOMtz8NOQmMdIxgpmcGU10T6GsbQEdLGxJOArlSjvLqchRWFCG3NAf5FQWiiCdBCoCFvgV87FrD27Yl3K3c4WzuBDsTOxjrGEMmlUHC%2FW9iowKgUMpRJa9CXnk%2BskqykFaYhhcFL5Ccl4InOU%2FwKPsRcspyoVKx5EhihwMHM30zeFi5w72BO1wtXeBk7oRGpo1ga2wDfW19SDnpP7ahUqlQWl2K1PxUJGYn4X5GHG69jEZseiwq5ZXC8ocQUoJxHIdm1l4Y5DEQvV17w8vaC1oSKbH2lSolskqyEJ%2F5APcz7uNWajRuvYxGcWUxGzEieKp7WXuiq9N7aOvQFt623nAwsf%2FLQ4IURRVFuJRyGccSIhD5KBIlVaVMAOqCkY4RRnv7Y1zrsfCy9lTrtSvllbiRegPhCREITziGvLI8NpoERGu71hjZYgQGuPeHnYmd2q9fXFmMI%2FFHsSN6FxJfJzIBqAmW%2BhaY2nEqPm47EcY6xhq3p1JeiYjE49h2awfuvLrDRhel6Gjp4IPmwzGl3WS0aNicCptUKhVOPz6DVZfWIDY9lgnAP6En08O0DkGY0Wk6jHSMqHOWCipcfX4Nqy6uwbXn19iIowSZVIaxrT7C7C6zYa%2BBp%2F27vmaGPgzDkqhleFnwkgnAm%2FR17YNV%2FVaisVkj6gNOpVIh8vEpLDqzGE9zn7ERqEF6Nu2B5X2%2BhruVmyDsLa0qxZpL%2F8H3N7eiWlHNBMBMzwyr%2B63EiBYfgAMnqOCrkFdg%2FZUN2Hh1M6oUVWw0qvM10cASq%2FquxPBmQ8FxnODsv5t%2BD0GhU5GU%2Faj%2BCkCnRh2xY%2Fg22BnbCToYY9PvIyA0UOM3s77Q26UXvh28CdaG1oLuR3l1OeafWoCfY%2FZABc0MQ6mun94SdV9Uwkkwq8tMbB2yBSa6JoIPyIZGDeHvPQqZxZmIz3zARihPaEmkWNRzEda%2Bv5rKNaLarF30desDZ3MnnHt2XiOvBGqfARhqG2Lr0C0Y6DFAlEG6I3onvjz1FeRKORuxBDHVNcEPI3aje5NuouxfXGY8xhwYi1Q1LxCqVQDsTOyw338vNZ9o%2BOLskyhMOvIxiitL2MglgIOpAw6NPgCPBu6i7mdWSRZG7x%2BLmLQY9c3G1XUhL2tPnP34lOgHPwD0cumJsHEhsNC3YKO3jrhauiJy4nHRD34AsDa0RsT4MPR17SMuAWjn4IsTEyNga2xbbwLXx84H4eNDYWlgyUZxLXGzckPEhDDYm9jXmz7ra%2Bvjl1E%2F48MWI8QhAO0cfHFkzGGY6prWuwBuZu2FsLFHYaZnykZzDXE2d0L4uBDBr%2FTXBplEhq1DtmB482HCFgCPBu44MHofjHQM620gN2vYDAdHH4C%2BTJ%2BN6hpMhUPGHkVDo4b11gdSiRRbh2xBN2c%2FYQqApYEl9vvvg5meWb0PaF%2BHttgxfBukBE8winkKfGD0XkHsCOUbbak2fhyxG00tmgpLAKScFDuGbWM38U8McO%2BPhd2%2FYo74p2DkJNgy%2BFu0sm3FnPE7pnqm%2BPnDH6En0xOOAMzoNF2032vrQnCn6RjWbChzxFvjZhqGeg1hjngDL2tPLO%2B9TBgC0MzaC%2FO7zWV37W%2FgOA6bBm6Ai6ULc8YbdGzUEV91X8Ac8RYmtZmIbk386BYAqUSKTYM2Qluqze7YWzDSMcKu4duho6XDnPE7Znpm2DFsK7QkWswZ%2F%2FDwWN9%2FHfFXAaIeH9d6LHzsWmvMSUqVEi8KUpH0OgnP858jszgLhRWFqJRXguM46Mn0YKZnBhujhnAyd4KblStsjWzVfpqspU1LzPObi6VRy1hkA1jTb5VGvvW%2FLsnGg6wHSMp%2BhJS8FGQWZyK%2FPB%2Fl1eVQqVTQ0dKBia4JrI2s0disMTwauMPbxhsNjTTzadLJvDGCO03HqotryAkLqa3AxrrGiJkeDSsDK7U6JbcsFycfReLs4yhcT72B3NLcdz5ZxYGDlaEVfB3aws%2FZD31d%2B6gtoYRcKUffH97HnVcx9Xrw93d%2FH7%2BO2qOWo%2BDVimpce3EdJ5JO4sKzi0jOS4ZSpazZlJmToKlFU%2FRy6YkhXoPhY9eal%2FyCb6O0qhRtvm2HjOIMugRgrt8XmO%2Bnnnd%2FFVS4%2FvwGdkTvxKnHp4llYpVKpGjn0A7jW4%2FFEK%2FBvE%2FTH2Q9RLcdPahIDKGp16FbU6%2FzvkM0rSgdu6J3Y%2F%2F9A8gsziTatruVOyb7fgJ%2F71G8rdS%2Fyc7oXZhzci49AmCqZ4r7wXfVcrT3ZuotLDv3Na6%2FuMHrdWyNbRDcKRgTfMbxKgRfnV6I7258Xy8FYHnvZZjWcSpv7WeVZGH1xbX49d5e3hO22BrbYK7fXIzxHs37fo8KeSV8NrdFWlFa3R96JPIBBLYPQB%2FX3rx2Or88H7OPf4b5pxeoJadacWUJop5GIexhOFwsXeBk3piX67RxaIu9sftQKoAU0iRxsXTB1iFbeBksSpUSO6N3YezBcbj58hYUKoVa4uXUo1M4%2F%2Bw8fB18eT0DoiXRglKlwIVnFzUvANpSbez6YAev232vv7iBIXuG4UbqTbUHal55Hg7FH0ZOWQ7ec%2BpCfKVaR0sHhtqGOP34TL0SgG8HbYI7Dyf8skqyMObgOOy6vRuVGkjTll6Ugb2x%2B2Bt2AAtbVrwdh13Kzfsit5d55lNnVcv%2Bru%2FDxse92z%2FHLMHg%2FcMRXpRusaCVaVSYVf0bvT%2FcRDxd0gAGNPqI163e9KGr0NbvO%2FWj3i79zPi0H1HL1xKvqzR%2FpVXl2PGsZlYcPor3mYfJromGNHigzq3U2cBGO8zjjdHbry6GTOPz6ZmkSwmLQZ9f%2BiPlLwUou3KpDJ80fXzeiMA8%2FzmEv%2F0eiP1Jgb9PITIezGRhwZU2HJjK4KPzeJNBMa2GqNZAbAzscN7Tl146dyOWzuxNGoZdfX6nuc%2Fx%2BA9Q%2FGy8BXRdoc1G4omFk1EP%2Fjb2PsQP%2BF2N%2F0eRu71R2FFIXX9%2FfXeXnx5ip8zIK3tWsHNylVzAjDUczAv30DPPDmL%2BacXaCxT6r%2BRWvASI%2Ff6E60dqCXRwtQOgaIXgOkdpxF9%2Br8qTIP%2Fvo9QVFlEbZ%2B339qBH%2B78yEvbdT07UafRO8hzEPEOpRWmYUpIIPWllxNeJ2Bq%2BHSiIjWq5UhY6JuLdvA7mjqiv%2Fv7xNqrVlRjwuFJyCrJor7v808t4CVj9CCPgZoRACsDK%2BLbflUqFYIjZiG%2FPF8QAX0sIQJ7Yn4h1p6%2BTB8ftRotWgGY6DOe6FeUtZfXCaZWY6W8ElPDp6OacLZoD2sPOJo6ql8AujfpRvwb7rHECEQ9PSeooF50dgmxbZnAb%2BcphFjp5l1ecfy9RxFrL%2FF1IjZe3SQoH8RlxOGnOz8RbZMDhx5Nu6tfALo4dSbakWpFNb4%2Bt1xwgV1YUYilUV8Ta6%2BpRVP4OviKTgC6N%2BlGLMWXCiosOL1IkOXY1l5eR3zTV5fGndUrABw4dGzUkWgnwhOOCbbQ5uG4I3iYlUCsvRHNh4tOAD4g2Kerz6%2Fh%2FLPzgvTD65LX2Be7n2ib7R3b1XoxvlZ%2FZWlgSTzd147oXYINboVKgfVXNhBrb6DHAFHlD9TR0iGa63795Q2C9sfuOz8S%2FbxtY2wDW2Mb9QmAt21Lop%2F%2FnuY%2BFcxizj%2BtX7witDfA2tBao3kVSNOpUUcY6xoTe%2Fe%2FlHJZ0P5Iep2E%2BxlxRNcBWtq0VJ8AeDbwJOqQE0kna3wumzaqFdVEp3a9XHqKRgB6EuzLL%2Ff2Cj5WAODko5NE2%2FOy9lSfALg3cCNq%2FPlnF0QR6IfijhCb2nV16ioaASC1W1SulCP0QagofHIx%2BRLR9tys3NQnAM7mzsQMlyvliEm7K4qb%2BjT3KRJek1kMbGnbEobaBoL3ibmeGTysyJz6i3kVgwweDmNpgriMeGKJbADAycxJfQLQqA4bD94kJe85SkRURffMk7NE2tGRasPb1lvw%2FvCx9yG2oHn2aZRo4qRCXoFnBL96OZg6qEcAdLV0iVb7Sc5LhpggeRS1jb2P4P3Rxo5cHy6nXBVVrKTkkztVaqJrXKvyc5LaXIhkiiySu%2Bho4E5aDLFzDGKokEOqD5XyStzPuC%2BqWMksJneGQVuqDZNafGmphQCQrXSbX14gqptaUlmCJ7lPyawD1PLTDi1wHIfmNs2JtJWUnUT0nZkGCivInmCszdissQAY6xoRNVpsNxUAEgjtCnQwsRd0aXErAytYGzYg49PXiaKLk2rCW5kNa5GWr8YCoCMlmyFXIsKDL6S2NEslUrhaugrWD%2B5WbsQ2jD3LTRZdnEgI7%2FbUqUVFLommja7NwgXtpBakkhtEPCTOVJ8AkLP9ZeFL0cUJ6c%2B8tRHbGv%2BFivAuLHMRJsAgubBZ2w0eNFDXdFV%2FJqs4S3RxYqFvQbS92uQerLEAkH5n5zOjsKbILcsj1lZTAecJJJnjkKRPacGmlgd43kZFLcZmjQWghPBZ5tpuYKAZkrkCm5gzAQCAEoI%2BpQVHghvqauujGgtAAeF0XfYm9qIrJ15eXU6sLTsTW0H6R0%2Bmh4aG5KroVojsa5GeTI94TcTafFKvuQBUFBLNxKKrpQsncydR3VyFUk4sWaiulq7GylHXBVtjW8ikMiJtqaBSS3kvdeJq6QIZwfyIFfIKFFUW8i8A5dXlyCP8PsZnCSVNwHESouWuHU0bCc4HJM%2BLcODUWoJbHZA%2B55Fdko1KeRX%2FAgAAz%2FNfEDXe176tqG6ujHD9QAdTe8H5wN6ErM3ahGYTtNCOcN7HlPzntfq7WgnAk5wnRI3v4tSZ6BNT0%2Bhr61M9mIQoAPoyA9HEh4ST1CmR59%2FxOPux%2BgSAZALM396HXNHIrJFobrCxjgnZwWRsJ0ABIGuzqZ6JaOLDs4EH8a9fD7IeqE8A7qXfI%2FzOzGGQxwDR3GDS1X1IrxarA9I2k940o0mG1LGc199xN%2B2e%2BgQgLjOe6KcuABjZcqRoCmI0JLy5yU6AAkDaZhsjG1HEhpZECx%2B2GEG0zcKKQiRmJ6lPAMqry3EnLYZoJ7ysPdHBsYMobrIj4emdrYmwBIADBxvCAuBo5iiK2Ojr1od4fNxIvYlqRbX6BAAAzj0lX5hhZucZorjJpMt8G%2BsYw0jHSDD9N9UzhQHhQ14uFk0FHxccx2FW52Di7UY9qX05vVoLwMmkSOId6dW0Jzo4thf8jfZs4EG0PQknEdQU2NbYhvjrnKe1p%2BDjYqDHAPjYkU3zplQpEfnolPoF4EnOEzzIekhcIVf1WynoqjjGOsbEZwAA%2BVV1PuHjs6WjqQMsDSwFGxcG2gZY3vtr4u3eTL2FtKI09QuACiocuH%2BQeIda2rTAzE7CfRVo5%2BjLy641Ie0FcDAhf8Drt3qUwl0jWt57GfF3fwDYX8diNHWK1AOxB3hJ6TXPb65gb3a3Jt14aVdI%2ByQa8bRg16NpD0HGxIjmH2BCm%2FHE2y2oKETIw1DNCUBOWS4Oxx8h3jGZVIafP%2FwRzgI7JCThJHjfrS8vbTcWkAA0NmvMS7v9XPtAJhHWluAOjdpj86CNvOx0%2FTlmD0qryjQnAACw%2Bdq3xNJg%2FxkrAyuEjD1K%2FMw0n%2Fg6%2BPIW%2FELKC9CEpyQmDQwboHvTbgKKh7Y44L8PejI94m2XV5dj681tdX9o1bWBxzlPeJkF%2FPHUOzkxotaFD9XNJ20n8TionAWxUUpLolXrMlXvwmTfTwURC31c%2ByBk7BGY6PKzhXnX7d3IJFAmjchq1YoLK1FWXcZLR%2B1N7HFq0kkM9hxE%2FVOPTxuNdIwEsRDoZN4YejJd3trv1sSP6opJUk6KOV0%2Fxz7%2FX2CobcjLNXJKc7DuygYyr60kGnlZ8BLrCRn0tuD%2F6cMfsHnQRt4Uta4s7rmQWAKMt%2BHVgP6ZkFcDL17bl3ASLO%2B9jMr8AC6WTRExIRwLus2HlOPvU%2FbiqKUoIFRQh5gXN1%2F7DvGZD3jrNAcO41qPxY2ga%2FiwxQiqAmCgxwAMVMNhJiFUCvK25d%2FG9o7t8anvJ9T02VjXGIt6fIUrAZd4%2F3p17ul57I89QE5QSTVUpahCQGgQyqsreHWArbENdgzbhguTo%2FC%2BWz%2BNC4GzuRM2DdyglnwGQigW2lZNyV2W9VqC9o7tNNpXE10TzOo8E3en38HsLrOgq6XL6%2FVySnMw%2FdgMKAmm5udMFpurSBo5wWc8Ng5cr7abkJSdhO23duJI%2FFGi2XjfBRsjG5yYeAzO5s5quV5BRQGarnGDXCmncvDryfTw7Isn0Odh1fvvyC%2FPx5A9w3A%2FI06t%2FfRo4I7xPuMxuuUoGNeiIGdtUKgUGLnXH1FPzxFtl7gAAMDGgRswwWecWm9KSWUJjiUex%2BH4w7iScpX3QeJl7YV9o35R%2Bwadnrt6486rGCoFoKtzV4SPC1HrNYsrixEQGoQTSSd5nnnaYrDnIHzQfDha27VSewarJWeXYuO1zeRfrfkQAJlUhoOj96N7E818s80ty8Xpx2dw5slZXEq%2BjHyCqcy1pdqY0m4yvuw2j5fvu%2F%2FGmktr8c2FVVQKwDd9VyCofYDar6tSqbAvdj%2BWnV9OrIKQTCJDS5sW6OHSA31cesHb1ltjr5s%2FxfyMmRGzeWmbFwEAflu5Dx17VOPvrdXKasRlxOPq82u4%2FuI6YtLuIqc0p8btmOmZYUTz4QhsH6DRNOYJWQnotPU9YmnHSSHlpLgXfEejG7fKqsqw%2F%2F4B7I3dj9j02Bq9KxtoG6B5w2Zo5%2BCLDo06oL1DO5hSUJk59GEYPj06hbcZLW8C8MegCRl7GK1sW1ETqAqlAs%2FzX%2BB%2Bxn0kZT9Ccl4y0grTkFuWh9KqUihVCuho6cJC3xz2JvbwtPZEB8f2aOfgCx0tHY3br4IKXbZ2JX4Ss650bNQBJycep8aejOIMRL%2B8jYdZCXhV%2BAp5ZXmokFdCyklgoGMAC30L2BnbwtncGe4N3NDUoil1BViOJUTgk6OTidbhUKsAAICprgn2j94ninP%2BtLDt1nbMi%2FySLpuGfo9RLUeym0OIA%2FcPYvqx4Fpn%2BnlXeH%2BpKagoxLBfPuB9kaY%2BMbqlP8z0zKixx9HUEUN5SHRZX9l4dROCwqbxPvgBQKrrp7eE74vIlXKEPQyHnkwfvg5tRZP8U1PoaOlAR0ubl7RstWHTwA1o1tCL3Zg6UiGvQHDELHx7%2FTu1rfGobVlTqVJi0dnF%2BPjIJ2r%2FXi9GJvt%2BCl%2FC1WVqwxDPwRjsNYjdkDryoiAV%2FX4YgL339qn1umr%2FrhHyMAw9dvZCbPp9dtfrMnWTSLFr%2BHaN5stvatEEGwdtEFVVJ00Q%2BjAMftu7E6%2B3QaUAAL8dIe7zQz%2Bsv7KR2l1tQsDR1BF7PvyR9y2of4elgSX2%2B%2B%2BDqa4JuxG1JL%2B8AAGhQZh0%2BBOie1WoWwP4OxRKBS6lXMa5p%2BfR2q4VrA2tWUTUUgTcrdwRkXSc6B7xf8JUzxRHxxwWTJ4G2lBBhWMJEfDfPxo3U29pdiapKQH4g4ziDPxy91cUVhShjX0b6FLwrV1ouFq5wqOBO04%2BiuQlO9ObT%2F6QMUfQinB56%2FrCk5wnCAgNwrorG1BSVaJxe3jfB1ATGhg2wJfd5mFMq4%2BgRbjEdn3gYvIljDs0AUUVRby039isMQ5%2FdAAuli7M2TXkjyQeu2%2F%2FwOvGHkELwH%2BfaJau%2BLLbPAz0GCDoGgGaICErAf4HxuBF%2Fgui7XZwbI89I3%2BClYEVc3INKKwoxLab2%2FH9za0o5EmYRScAf%2BBl7YXZXWZisOcgNiOoydOmLBefHPkUF5Mv1T1AwGFS24lY2XcFdVtlaX%2Fib4%2FeiV3RuzW2wCd4AfgDJ3MnTGk3GR95%2BwuqRp4mkSvlWHVxNTZc2QSFqnbrAkY6RljXfy3xarZi5lH2Y2y%2FtQMH4w7WOWU3E4A3MNY1xsgWH2KCz3i2Av2OXEi%2BiMDQqTXOINvarhV2DNuOphZNmBP%2FhUp5JU4%2BisSPd37CledXoVIJZkgJSwD%2BPC1tZeeN0d7%2BGOI5WNA149RBblkuZh%2F%2FHOEJx%2F71d7UkWpjdZSbmvPc570lOhYxSpURMWgwOxR3B0QchyCvLE2Q%2FBCkAf0YmlcHPuSsGeQxEX7e%2BsGJi8LeooMKR%2BKP44uS8t76Tulm54fsh38HHrjVz2Fteq2JexeB40gkcSzxOfKGVCUBdxUAig499a%2FR26YXuTbujmbUXWzx8g4ziTMw5%2BQWOJ574i4jO6DgNc7p%2BrpFdhTSTWZyJyylXcO7ZeZx%2FegHZpdmi6p%2BoBOBNLPQt0LFRh98Sejj6wsvaiwX477OBsIfhmBc5H7bGttg4cANa2rRgT3ilHM9ynyEm7S5upUbj5subeJLzVG07LJkA8IyOlg48GrjDy9oLXtZecLV0QROLJrA3sRNc0UkSFFUUQV9bv97Nkoori%2FGqMA0v8l8gOS8Zj3MeI%2BF1IhJfJ6K4sqRe%2BaJeCcDbkEqksDa0ho1RQ1gYWMJU1wSG2obQ1tKGhJNAAu6%2FOQy4P%2F03wEHyx79zEkg5CaQSKWRSbWhLZdCT6UFfpg9jHWOY6pnC0sASDQ2tqcg1RxNKlRLZpdlILXiJjKIMvC7NRkF5AYoqi1BWVYZyeTmqFVWoUlRDoVRAoVICKhWUUOKPY%2FMqqP67%2Bv7n%2F1YoFahUVKG0qgQF5QXIKctFaVUpc%2FrvsBfk34MkvSgd6UXpvF9LwklgomuCJhbOaGbthbb2bdHZqTMaCagKcl0pry7HtRfXcSn5Mu68uoOHrxNQXFksqM9nbAbAICoKXtae%2BLDFh%2FjI2x%2Fm%2Buai7GdcRhx23t6NYwkRKKwoZDeeCQDjTQx1DDHZ91PM6hwsml2PyXkpWHhmMSIfRYp6QY0J
```

**Comment by ssborbis** [comment]
> A victory in the night! No issue with exporting. I imported it multiple times. 30,4 MB if I click on 'Delete search engine icon cache' first, 64,2 MB if I don't.

Good deal. You must have hit some obscure DOM limit with that large export.

>I recently tried the Context Menu after the external apps feature, so I used those fast buttons. Maybe I miss a piece but from the following image shouldn't the Context Menu must show all the engines selected with 'A' when I perform a selection.

It *should*. I'll do some testing locally. 

> To preserving the .ico file, I always use ACSE extension to convert the icon. Then I paste the code into CS 'Icon URI'.

I'm curious, why don't you simply upload the image through the edit modal, and let this addon convert it to a data uri?

![Screenshot_2022-03-02_22-30-16](https://user-images.githubusercontent.com/24310997/156496722-245ab2ec-28e7-48ef-b500-580d6d45167a.jpg)


> Maybe that's why I receive the metioned error if I try to export the search engine from CS to Firefox (obviously I don't if I use CS's icons selected or converted):

Hmm... Your larger icons may be running into the upper limit of what the OpenSearch API will tolerate. I think that limit may be set by Firefox, and I don't know what it is. But I do recall another issue years ago by another users. It turned out to be a large image.

**Comment by logusergithub** [comment]
> Good deal. You must have hit some obscure DOM limit with that large export.

Good has won. Not to repeat myself, I already consider a miracle managing so many search engines with those icons without slowdowns. The foundations of CS are solid.

> It should. I'll do some testing locally.

In this case I still have to try with a new profile to exclude some personal config issue.

> I'm curious, why don't you simply upload the image through the edit modal, and let this addon convert it to a data uri?

I make a backup with all of the icons I personally made and used. Moreover many search engines you have seen aren't in Mycroft Project or have decent icons available.
The result of the edit modal is the image above (Statcounter.png after I upoaded Statcounter.ico). When I do a backup with CS I also consider it as a second backup for icons. I will use edit modal if in the future it will preserve the icon files during conversion.

> Your larger icons may be running into the upper limit of what the OpenSearch API will tolerate.

Edit modal preserves the size 256x256 of the icon during conversion to data URI. Moreover if I use the ACSE extension (it does the same work of edit modal to data URI) to export search engine with my icons I don't encounter any issue.
Maybe I miss something, but you can try yourself:
- copy from one of my search engine the 'Icon URI' code to adress bar and save the icon file
- use edit modal to import the icon saved
- copy new 'Icon URI' code to address bar and save the resulted png
- see the size (obviously without resize the original 256x256)

I suppose there is a incompatibility with the data URI from ACSE using CS to export the search engine. But I'm not consider it a issue.

**Comment by logusergithub** [comment]
> In this case I still have to try with a new profile to exclude some personal config issue.

On new profile, new CS 1.36 installed, "Filter search engines based on the target element" and "Remove subfolders and flatten menu if only n number of engines are to be shown (0): from what I see there is some issues related to saved settings from fast buttons.  At the most to better explain I can make a video.

**Comment by ssborbis** [comment]
I have yet to run into this problem. I wonder if the number of engines you have is causing problems. The context menu is a bit odd in that you cannot make it wait for the menu to be built before opening; it will display ASAP whether you want it to or not. So if you have a large menu to build, it will simply display the last menu built and not update once the new menu is finished building. It's a major pain, honestly.

Now that you can export that monstrosity of a config of yours, can you get me another copy to test with?

**Comment by logusergithub** [comment]
> Now that you can export that monstrosity of a config of yours, can you get me another copy to test with?

Let me finish to modify some old engines with {domain} parameter, add new engines, modify {url} parameter for some app launchers and I'll post it if necessary. Above I used a new profile and a new installation of CS on Firefox, every time I made a change I restarted the profile and tried a context 'selection' search:

- when you start with the above settings mentioned
https://i.postimg.cc/s2BXkWHk/2022-03-06-011652.png

- I removed all the 'A' with fast buttons
https://i.postimg.cc/Y96p7z6s/2022-03-06-011932.png

- I deselected 'Wikipedia it' (from Firefox) using fast button and deselected\selected 'Google Site Search' using the engine settings windows
https://i.postimg.cc/mkVBzZ63/2022-03-06-012119.png

- I deselected\selected 'eBay' using the engine settings windows and deselected 'Google Site Search' using fast button
https://i.postimg.cc/QNS3qt01/2022-03-06-012552.png

Maybe a video is better, but I feel there is something wrong with fast buttons. Tell me if you still need my 'monstrosity of a config', I'll post it tomorrow night. This became my favourite extension and I'm happy to contribute.

**Comment by ssborbis** [comment]
I think maybe some of the problem is - if you don't have any contexts selected, it will default to "Selection" so the A will keep coming back. I'll look into this further.

**Comment by logusergithub** [comment]
> f you don't have any contexts selected, it will default to "Selection"

Yes, but only in appearance, not when you perform a 'selection' search. It seems that fast buttons are not saved. Context options from engine's window works perfectly. If I have to explain you from the point of view of a a biotechnologist seems that fast buttons are 'recessive' and engine's windows are 'dominant' like alleles (a fun crossover explanaton).
Moreover on my 'monstrosity of a config the question is: why do I see a mixed result? In my opinion the correct functionality was lost in some subsequent update.

**Comment by ssborbis** [comment]
Ok, I see what's going on. 

I have code reading the contexts from different keys. Some are reading the shortcuts, some are reading the true engines.

I have a fix. Thanks for the help. Fixes going up to the dev branch now.

**Comment by ssborbis** [comment]
v1.37 going up. See if that fixes things.

**Comment by logusergithub** [comment]
At a guess it seems that 1.37 fixed all the issues mentioned in these days! Thanks.
The first thing I noticed is that all quick button previously edited and saved are matching with no mixed results in the context menu.
